### ISSUE TITLE
Zone information not available on some clusters

### ISSUE THREAD

**Issue by tgdfool2** [rook/rook#12139]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
After having migrated our clusters from "directory-based OSDs" to "device-based OSDs" (as described under https://rook.io/docs/rook/v1.3/ceph-upgrade.html#converting-legacy-osds), the`zone` property is not properly taken in account on some clusters.

Example of cluster where working as expected:
```
[root@rook-ceph-tools-689469b64c-xlzff /]# ceph osd tree
ID  CLASS WEIGHT  TYPE NAME                   STATUS REWEIGHT PRI-AFF
 -1       0.78119 root default
-11       0.78119     region north
 -6       0.39059         zone north-o
 -3       0.19530             host ab-c-n-o-0
  1   hdd 0.19530                 osd.1           up  1.00000 1.00000
 -7       0.19530             host ab-c-n-o-1
  0   hdd 0.19530                 osd.0           up  1.00000 1.00000
 -5       0.39059         zone north-z
 -9       0.19530             host ab-c-n-z-0
  2   hdd 0.19530                 osd.2           up  1.00000 1.00000
 -4       0.19530             host ab-c-n-z-1
  3   hdd 0.19530                 osd.3           up  1.00000 1.00000
```

Example of cluster where *not* working as expected:
```
[root@rook-ceph-tools-689469b64c-cc7mw /]# ceph osd tree
ID CLASS WEIGHT  TYPE NAME               STATUS REWEIGHT PRI-AFF
-1       0.78119 root default
-6       0.78119     region north
-2       0.19530         host de-f-n-o-0
 0   hdd 0.19530             osd.0           up  1.00000 1.00000
-3       0.19530         host de-f-n-o-1
 1   hdd 0.19530             osd.1           up  1.00000 1.00000
-9       0.19530         host de-f-n-z-0
 2   hdd 0.19530             osd.2           up  1.00000 1.00000
-4       0.19530         host de-f-n-z-1
 3   hdd 0.19530             osd.3           up  1.00000 1.00000
```

Nodes on both clusters are labelled with the same labels, e.g.:
```
# Running Kubernetes version 1.17.17, so not yet using the "new" topology labels
failure-domain.beta.kubernetes.io/region=north
failure-domain.beta.kubernetes.io/zone=north-o  (or north-z)
```

**Expected behavior:**
Zone correctly displayed for all clusters.

**How to reproduce it (minimal and precise):**

Migrate from "directory-based OSDs" to "device-based OSDs" as described under https://rook.io/docs/rook/v1.3/ceph-upgrade.html#converting-legacy-osds, remove and re-add all nodes (one-by-one) from the `CephCluster`

**Environment**:
* OS (e.g. from /etc/os-release): CentOS Linux 7
* Kernel (e.g. `uname -a`): 3.10.0-1160.el7.x86_64
* Cloud provider or hardware configuration: bare-metal
* Rook version (use `rook version` inside of a Rook Pod): 1.2.7
* Storage backend version (e.g. for ceph do `ceph -v`): 14.2.8
* Kubernetes version (use `kubectl version`): 1.17.17
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): from scratch
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): HEALTH_OK

Could you please help us to find out why some clusters are not displaying the `zone` information as expected? Goal of this issue is to ensure that our clusters are in a stable state before proceeding with the upgrade to Rook 1.3.x.

**Comment by tgdfool2** [comment]
Please find here a extract from the logs of the `osd-prepare` Pod where the `zone` information is not properly displayed:
```
❯ k logs rook-ceph-osd-prepare-de-f-n-z-1-9nbq7
Defaulted container "provision" out of: provision, copy-bins (init)
2023-04-13 08:16:04.510899 I | cephcmd: desired devices to configure osds: [{Name:sdb OSDsPerDevice:1 MetadataDevice: DatabaseSizeMB:0 DeviceClass: IsFilter:false IsDevicePathFilter:false}]
2023-04-13 08:16:04.513912 I | rookcmd: starting Rook v1.2.7 with arguments '/rook/rook ceph osd provision'
2023-04-13 08:16:04.513943 I | rookcmd: flag values: --cluster-id=xxx, --data-device-filter=, --data-device-path-filter=, --data-devices=sdb:1:::, --data-directories=, --encrypted-device=false, --force-format=false, --help=false, --location=, --log-flush-frequency=5s, --log-level=IN
FO, --metadata-device=, --node-name=de-f-n-z-1, --operator-image=, --osd-database-size=0, --osd-journal-size=5120, --osd-store=, --osd-wal-size=576, --osds-per-device=1, --pvc-backed-osd=false, --service-account=
2023-04-13 08:16:04.513954 I | op-mon: parsing mon endpoints: ax=1.2.3.4:6789,av=1.2.3.5.6789,aw=1.2.3.6:6789
2023-04-13 08:16:04.534714 I | op-osd: CRUSH location=root=default host=de-f-n-z-1 region=north zone=north-z
2023-04-13 08:16:04.534738 I | cephcmd: crush location of osd: root=default host=de-f-n-z-1 region=north zone=north-z
2023-04-13 08:16:04.546135 I | cephconfig: writing config file /var/lib/rook/rook-block/rook-block.config
2023-04-13 08:16:04.546320 I | cephconfig: generated admin config in /var/lib/rook/rook-block
2023-04-13 08:16:04.546465 I | cephosd: discovering hardware
...
```

Looks like the `zone` information is available in this log file; so why not properly displayed in the `ceph osd tree` output?

**Comment by sp98** [comment]
@tgdfool2 can you please provide the complete osd prepare logs.  Also the rook operator logs and OSD  pod yaml. Thanks.

**Comment by tgdfool2** [comment]
@sp98: requested information has been sent to your email address

**Comment by tgdfool2** [comment]
Sure, let me join the Slack channel

From: Santosh Pillai ***@***.***>
Date: Tuesday, 25 April 2023 at 15:11
To: rook/rook ***@***.***>
Cc: Fournier Olivier, INI-NET-VNC-TPN ***@***.***>, Mention ***@***.***>
Subject: Re: [rook/rook] Zone information not available on some clusters (Issue #12139)

It went to the spam folder and not allowing me to download it. Can you share via Rook slack channel using the same email ID. Thanks.

**Comment by tgdfool2** [comment]
Just for the records: after some troubleshooting, we found out that some clusters were missing the `zone` type in the crushmap:
```
# types
type 0 osd
type 1 host
type 2 chassis
type 3 rack
type 4 row
type 5 pdu
type 6 pod
type 7 room
type 8 datacenter
type 9 zone                   # <= this one was missing
type 10 region
type 11 root
```

The following actions have been performed in order to solve the issue:
* Export, decompile, adjust, recompile and reimport the crushmap (https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/1.2.3/html/storage_strategies/editing-a-crush-map#doc-wrapper)
* Create missing buckets and move OSDs to the newly-created buckets (https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/1.2.3/html/installation_guide_for_rhel_x86_64/create_a_crush_hierarchy)

Thanks again for your help @sp98.
